### PR TITLE
Fixed issue 3615 by updating schedule.filter() to accept scalar values as well as iterables

### LINKED ISSUE #3615: Schedule.filter should accept scalar values as well as iterables
### What is the expected enhancement?
`Schedule.filter` currently expects its input arguments to be a list. This should be extended to also accept a scalar value.

```python
schedule.filter(instruction_types=DelayInstruction)
```

### PR DESCRIPTION
**Pull request by NuoWenLei**
### Summary

Fixes #3615 

### Details and comments

In construct_filter:
Using try/except detecting TypeError for all parameters except channel, it can now pass scalar values as well as iterables

For Channel - in schedule.filter():
Passing only one channel resulted in no filter whatsoever in terms of channel, therefore in schedule.filter() when one channel is passed, the exact same channel is added to create an iterable for construct_filter().

### CODE REVIEW

**Comment by NuoWenLei** [comment]
Hi, what should I do with the 2 checks that did not pass? And is there any way to look closer as to where the errors occurred?

**Comment by ajavadia** [comment]
@NuoWenLei you can see the error by clicking on "Details" and then the red crosses.
In this case it says:

> qiskit/pulse/schedule.py:18:0: W0611: Unused Set imported from typing (unused-import)

**Review by taalexander — COMMENTED** [review]
Withing `_construct_filter` could you define a method `scalar_to_list` and just call that on the arguments of the generator methods

```python
def if_scalar_cast_to_list(input)
try:
    iter(input)
except TypeError, te:
    input = list(input)
return input
```
You would also have to create a`time_ranges` generator function and remove the set cast from channels. This may be the cleanest approach.

**Comment by NuoWenLei** [comment]
Hello, I have a question regarding my local qiskit. I've made the requested changes and I'm hoping to test it out with jupyter notebook. However, it seems that one of my files in qiskit-terra is messed up, and so it would not run. Is there any way of updating my local qiskit-terra branch without needing to make a pull request for all the new files? Sorry for such a simple question.

**Comment by NuoWenLei** [comment]
Ok, I'm working on it!

**Comment by NuoWenLei** [comment]
I've made the changes, how do I update the pull request?

**Comment by NuoWenLei** [comment]
This new pull request has the function asked by #3615

File: qiskit/pulse/schedule.py [hunk 1]
```diff
@@ -256,11 +256,18 @@ def filter(self, *filter_funcs: List[Callable],
             time_ranges: For example, [(0, 5), (6, 10)]
             intervals: For example, [Interval(0, 5), Interval(6, 10)]
         """
+        if channels is not None:
```
Comment (taalexander) [inline qiskit/pulse/schedule.py]: Is there a reason the logic is here and not `_construct_filter`?

File: qiskit/pulse/schedule.py [hunk 2]
```diff
@@ -324,34 +331,51 @@ def _construct_filter(self, *filter_funcs: List[Callable],
             time_ranges: For example, [(0, 5), (6, 10)]
             intervals: For example, [Interval(0, 5), Interval(6, 10)]
         """
-        def only_channels(channels: Set[Channel]) -> Callable:
+        def only_channels(channels: Channel) -> Callable:
```
Comment (taalexander) [inline qiskit/pulse/schedule.py]: I believe this is still accepting an iterable? Should this be a type of `Union[Set[Channel], Channel]`.

File: qiskit/pulse/schedule.py [hunk 3]
```diff
@@ -324,34 +331,51 @@ def _construct_filter(self, *filter_funcs: List[Callable],
...
             def channel_filter(time_inst: Tuple[int, 'Instruction']) -> bool:
                 return any([chan in channels for chan in time_inst[1].channels])
             return channel_filter
 
-        def only_instruction_types(types: Iterable[abc.ABCMeta]) -> Callable:
+        def only_instruction_types(types: Optional[Iterable[abc.ABCMeta]] = None) -> Callable:
```
Comment (taalexander) [inline qiskit/pulse/schedule.py]: Should this type be `Union[abc.ABCMeta, Iterable[abc.ABCMeta]]`

File: qiskit/pulse/schedule.py [hunk 4]
```diff
@@ -324,34 +331,51 @@ def _construct_filter(self, *filter_funcs: List[Callable],
...
-                return isinstance(time_inst[1], tuple(types))
+                try:
+                    iter(types)
+                    return isinstance(time_inst[1], tuple(types))
+                except TypeError:
+                    return isinstance(time_inst[1], types)
             return instruction_filter
 
-        def only_intervals(ranges: Iterable[Interval]) -> Callable:
+        def only_intervals(ranges: Interval) -> Callable:
```
Comment (taalexander) [inline qiskit/pulse/schedule.py]: Similar union type update here.

File: qiskit/pulse/schedule.py [hunk 5]
```diff
@@ -324,34 +331,51 @@ def _construct_filter(self, *filter_funcs: List[Callable],
...
             def interval_filter(time_inst: Tuple[int, 'Instruction']) -> bool:
-                for i in ranges:
-                    if all([(i.start <= ts.interval.shift(time_inst[0]).start
-                             and ts.interval.shift(time_inst[0]).stop <= i.stop)
+                try:
```
Comment (taalexander) [inline qiskit/pulse/schedule.py]: It seems we have the iterable check code here and down below prior to calling this on line 372. Is there a reason for this?

**Review by taalexander — CHANGES_REQUESTED** [review]
This looks much better! Couple of small suggestions.

**Comment by NuoWenLei** [comment]
Hi, I updated the union parameters in #3768 .

**Comment by taalexander** [comment]
Hello @NuoWenLei, could you please help me understand the difference between this PR and #3768?

**Comment by NuoWenLei** [comment]
Hi, sorry for the extra PR, I couldn't figure out how to update the changed code for this PR, so instead I created another PR and referenced this one. #3768 has the most recent and most complete patches.

**Comment by taalexander** [comment]
Ok, in that case, I will close this PR and review your other PR 😄.

**Pull request closed without merging**